File: t/CS/OmrProcessingDemo/IImageManager.cs
using Leadtools;
using Leadtools.Caching;
using Leadtools.Codecs;
using System;
using System.Collections.Generic;
using System.IO;

namespace OmrProcessingDemo
{
   public interface IImageManager
   {
      bool Has(string key);
      void Add(string key, RasterImage image);
      void Add(string key, string path);
      RasterImage Get(string key);
      RasterImage GetPage(string key, int page);
      void Clear();
      void Save();

      // no foreach or getcount
   }

   // This a test image manager, dont use it. only for debugging
   public class MemoryImageManager : IImageManager
   {
      private Dictionary<string, RasterImage> _images = new Dictionary<string, RasterImage>();
      private Dictionary<string, string> _imagePaths = new Dictionary<string, string>();

      public MemoryImageManager()
      {
      }

      public bool Has(string key)
      {
         return _images.ContainsKey(key) || _imagePaths.ContainsKey(key);
      }

      public void Add(string key, RasterImage image)
      {
         Remove(key);

         _images[key] = image != null ? image.Clone() : null;
      }

      public void Add(string key, string path)
      {
         Remove(key);

         _imagePaths[key] = path;
      }

      private void Remove(string key)
      {
         RasterImage oldImage = null;

         if (_images.ContainsKey(key))
         {
            oldImage = _images[key];
            _images.Remove(key);
         }

         if (_imagePaths.ContainsKey(key))
         {
            _imagePaths.Remove(key);
         }

         if (oldImage != null)
            oldImage.Dispose();
      }

      public RasterImage Get(string key)
      {
         if (_images.ContainsKey(key))
         {
            var image = _images[key];
            return image != null ? image.CloneAll() : null;
         }

         if (_imagePaths.ContainsKey(key))
         {
            if (File.Exists(_imagePaths[key]) == false)
            {
               return null;
            }

            using (var codecs = MainForm.GetRasterCodecs())
            {
               codecs.Options.Load.AllPages = true;
               RasterImage image = codecs.Load(_imagePaths[key]);
               image.ChangeViewPerspective(RasterViewPerspective.TopLeft);
               return image;
            }
         }

         return null;
      }

      public RasterImage GetPage(string key, int page)
      {
         if (_images.ContainsKey(key))
         {
            var image = _images[key];

            if (image != null)
            {
               image.Page = page;
               return image.Clone();
            }

            return null;
         }

         if (_imagePaths.ContainsKey(key))
         {
            if (File.Exists(_imagePaths[key]) == false)
            {
               return null;
            }

            using (var codecs = MainForm.GetRasterCodecs())
            {
               codecs.Options.Load.AllPages = true;
               RasterImage image = codecs.Load(_imagePaths[key], page);
               image.ChangeViewPerspective(RasterViewPerspective.TopLeft);
               return image;
            }
         }


         return null;
      }

      public void Clear()
      {
         foreach (var item in _images)
         {
            if (item.Value != null)
               item.Value.Dispose();
         }

         _images.Clear();
         _imagePaths.Clear();
      }

      public void Save()
      {
         // Not implemented
      }
   }

   public class CacheImageManager : IImageManager
   {
      private ObjectCache _cache;
      private string _region;
      private Dictionary<string, string> _imagePaths = new Dictionary<string, string>();

      public string CreateNewImageId()
      {
         return Guid.NewGuid().ToString();
      }

      public CacheImageManager()
      {
         // maybe change the demo to ask for the workspace directory before it starts?
         /*
         string workspaceDir = @"//myworkspace";
         string cacheDir = System.IO.Path.Combine(workspaceDir, "cache");
         CacheImageManager imageManager = new CacheImageManager(cacheDir, "OMR");
         */
      }

      public void Create(string directory, string region)
      {
         if (string.IsNullOrEmpty(region))
            throw new InvalidOperationException("region cannot be null or empty");

         var fileCache = new FileCache();
         fileCache.CacheDirectory = directory;
         _region = region;
         _cache = fileCache;
      }

      public bool Has(string key)
      {
         return (_cache != null && _cache.Contains(key, _region)) || _imagePaths.ContainsKey(key);
      }

      public void Add(string key, RasterImage image)
      {
         if (_cache == null)
            throw new InvalidOperationException("Must call create first");

         Remove(key);

         _cache.Add(key, image, new CacheItemPolicy(), _region);
      }

      public void Add(string key, string path)
      {
         Remove(key);

         _imagePaths[key] = path;
      }

      private void Remove(string key)
      {
         if (_imagePaths.ContainsKey(key))
         {
            _imagePaths.Remove(key);
         }
      }

      public RasterImage Get(string key)
      {
         if (_imagePaths.ContainsKey(key))
         {
            using (var codecs = MainForm.GetRasterCodecs())
            {
               var path = _imagePaths[key];
               RasterImage image = codecs.Load(path, 1);
               image.ChangeViewPerspective(RasterViewPerspective.TopLeft);
               return image;
            }
         }

         if (_cache != null)
            return _cache.Get<RasterImage>(key, _region);

         return null;
      }

      public RasterImage GetPage(string key, int page)
      {
         if (_imagePaths.ContainsKey(key))
         {
            using (var codecs = MainForm.GetRasterCodecs())
            {
               var path = _imagePaths[key];
               RasterImage image = codecs.Load(path, page);
               image.ChangeViewPerspective(RasterViewPerspective.TopLeft);
               return image;
            }
         }

         if (_cache != null)
         {
            RasterImage img = _cache.Get<RasterImage>(key, _region);
            img.Page = page;

            return img;
         }

         return null;
      }

      public void Clear()
      {
         if (_cache != null)
            _cache.DeleteRegion(_region);
         _imagePaths.Clear();
      }

      public void Save()
      {
         if (_cache == null)
            throw new InvalidOperationException("Must call create first");

         if (_imagePaths.Count == 0)
            return;

         using (var codecs = MainForm.GetRasterCodecs())
         {
            foreach (var item in _imagePaths)
            {
               using (var image = codecs.Load(item.Value, 1))
               {
                  image.ChangeViewPerspective(RasterViewPerspective.TopLeft);
                  _cache.Add(item.Key, image, new CacheItemPolicy(), _region);
               }

            }
         }
      }
   }
}

File: t/VB/OmrProcessingDemo/IImageManager.vb
Imports Leadtools
Imports Leadtools.Caching
Imports Leadtools.Codecs
Imports System
Imports System.Collections.Generic
Imports System.IO

Public Interface IImageManager
   Function Has(ByVal key As String) As Boolean
   Sub Add(ByVal key As String, ByVal image As RasterImage)
   Sub Add(ByVal key As String, ByVal path As String)
   Function [Get](ByVal key As String) As RasterImage
   Function GetPage(ByVal key As String, ByVal page As Integer) As RasterImage
   Sub Clear()
   Sub Save()
End Interface

Public Class MemoryImageManager
   Implements IImageManager

   Private _images As Dictionary(Of String, RasterImage) = New Dictionary(Of String, RasterImage)()
   Private _imagePaths As Dictionary(Of String, String) = New Dictionary(Of String, String)()

   Public Sub New()
   End Sub

   Public Function Has(ByVal key As String) As Boolean Implements IImageManager.Has
      Return _images.ContainsKey(key) OrElse _imagePaths.ContainsKey(key)
   End Function

   Public Sub Add(ByVal key As String, ByVal image As RasterImage) Implements IImageManager.Add
      Remove(key)
      _images(key) = If(image IsNot Nothing, image.Clone(), Nothing)
   End Sub

   Public Sub Add(ByVal key As String, ByVal path As String) Implements IImageManager.Add
      Remove(key)
      _imagePaths(key) = path
   End Sub

   Private Sub Remove(ByVal key As String)
      Dim oldImage As RasterImage = Nothing

      If _images.ContainsKey(key) Then
         oldImage = _images(key)
         _images.Remove(key)
      End If

      If _imagePaths.ContainsKey(key) Then
         _imagePaths.Remove(key)
      End If

      If oldImage IsNot Nothing Then oldImage.Dispose()
   End Sub

   Public Function [Get](ByVal key As String) As RasterImage Implements IImageManager.Get
      If _images.ContainsKey(key) Then
         Dim image As RasterImage = _images(key)
         Return If(image IsNot Nothing, image.CloneAll(), Nothing)
      End If

      If _imagePaths.ContainsKey(key) Then

         If File.Exists(_imagePaths(key)) = False Then
            Return Nothing
         End If

         Using codecs As RasterCodecs = MainForm.GetRasterCodecs()
            codecs.Options.Load.AllPages = True
            Dim image As RasterImage = codecs.Load(_imagePaths(key))
            image.ChangeViewPerspective(RasterViewPerspective.TopLeft)
            Return image
         End Using
      End If

      Return Nothing
   End Function

   Public Function GetPage(ByVal key As String, ByVal page As Integer) As RasterImage Implements IImageManager.GetPage
      If _images.ContainsKey(key) Then
         Dim image As RasterImage = _images(key)

         If image IsNot Nothing Then
            image.Page = page
            Return image.Clone()
         End If

         Return Nothing
      End If

      If _imagePaths.ContainsKey(key) Then

         If File.Exists(_imagePaths(key)) = False Then
            Return Nothing
         End If

         Using codecs As RasterCodecs = MainForm.GetRasterCodecs()
            codecs.Options.Load.AllPages = True
            Dim image As RasterImage = codecs.Load(_imagePaths(key), page)
            image.ChangeViewPerspective(RasterViewPerspective.TopLeft)
            Return image
         End Using
      End If

      Return Nothing
   End Function

   Public Sub Clear() Implements IImageManager.Clear
      For Each item As KeyValuePair(Of String, RasterImage) In _images
         If item.Value IsNot Nothing Then item.Value.Dispose()
      Next

      _images.Clear()
      _imagePaths.Clear()
   End Sub

   Public Sub Save() Implements IImageManager.Save
   End Sub
End Class

Public Class CacheImageManager
   Implements IImageManager

   Private _cache As ObjectCache
   Private _region As String
   Private _imagePaths As Dictionary(Of String, String) = New Dictionary(Of String, String)()

   Public Function CreateNewImageId() As String
      Return Guid.NewGuid().ToString()
   End Function

   Public Sub New()
   End Sub

   Public Sub Create(ByVal directory As String, ByVal region As String)
      If String.IsNullOrEmpty(region) Then Throw New InvalidOperationException("region cannot be null or empty")
      Dim fileCache As FileCache = New FileCache()
      fileCache.CacheDirectory = directory
      _region = region
      _cache = fileCache
   End Sub

   Public Function Has(ByVal key As String) As Boolean Implements IImageManager.Has
      Return (_cache IsNot Nothing AndAlso _cache.Contains(key, _region)) OrElse _imagePaths.ContainsKey(key)
   End Function

   Public Sub Add(ByVal key As String, ByVal image As RasterImage) Implements IImageManager.Add
      If _cache Is Nothing Then Throw New InvalidOperationException("Must call create first")
      Remove(key)
      _cache.Add(key, image, New CacheItemPolicy(), _region)
   End Sub

   Public Sub Add(ByVal key As String, ByVal path As String) Implements IImageManager.Add
      Remove(key)
      _imagePaths(key) = path
   End Sub

   Private Sub Remove(ByVal key As String)
      If _imagePaths.ContainsKey(key) Then
         _imagePaths.Remove(key)
      End If
   End Sub

   Public Function [Get](ByVal key As String) As RasterImage Implements IImageManager.Get
      If _imagePaths.ContainsKey(key) Then

         Using codecs As RasterCodecs = MainForm.GetRasterCodecs()
            Dim path As String = _imagePaths(key)
            Dim image As RasterImage = codecs.Load(path, 1)
            image.ChangeViewPerspective(RasterViewPerspective.TopLeft)
            Return image
         End Using
      End If

      If _cache IsNot Nothing Then Return _cache.[Get](Of RasterImage)(key, _region)
      Return Nothing
   End Function

   Public Function GetPage(ByVal key As String, ByVal page As Integer) As RasterImage Implements IImageManager.GetPage
      If _imagePaths.ContainsKey(key) Then

         Using codecs As RasterCodecs = MainForm.GetRasterCodecs()
            Dim path As String = _imagePaths(key)
            Dim image As RasterImage = codecs.Load(path, page)
            image.ChangeViewPerspective(RasterViewPerspective.TopLeft)
            Return image
         End Using
      End If

      If _cache IsNot Nothing Then
         Dim img As RasterImage = _cache.[Get](Of RasterImage)(key, _region)
         img.Page = page
         Return img
      End If

      Return Nothing
   End Function

   Public Sub Clear() Implements IImageManager.Clear
      If _cache IsNot Nothing Then _cache.DeleteRegion(_region)
      _imagePaths.Clear()
   End Sub

   Public Sub Save() Implements IImageManager.Save
      If _cache Is Nothing Then Throw New InvalidOperationException("Must call create first")
      If _imagePaths.Count = 0 Then Return

      Using codecs As RasterCodecs = MainForm.GetRasterCodecs()

         For Each item As KeyValuePair(Of String, String) In _imagePaths

            Using image As RasterImage = codecs.Load(item.Value, 1)
               image.ChangeViewPerspective(RasterViewPerspective.TopLeft)
               _cache.Add(item.Key, image, New CacheItemPolicy(), _region)
            End Using
         Next
      End Using
   End Sub
End Class
